group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeAssignor.java

```diff
@@ -0,0 +1,261 @@
+                } else {
+                    // 2) If remaining = 0: it has min req partitions but there is scope for getting an extra partition later on, so it is a potential unfilled member.
+                    // 3) If remaining > 0: it doesn't even have the min required partitions, and it definitely is unfilled, so it should be added to potentiallyUnfilledMembers.
+                    RemainingAssignmentsForMember<String, Integer> newPair = new RemainingAssignmentsForMember<>(memberId, remaining);
+                    potentiallyUnfilledMembers.add(newPair);
+                }
+            }
+
+            // Step 3
+            // If remaining > 0 after increasing the required quota due to the extra partition, add potentially unfilled member to the unfilled members list.
+            for (RemainingAssignmentsForMember<String, Integer> pair : potentiallyUnfilledMembers) {
+                String memberId = pair.memberId();
+                Integer remaining = pair.remaining();
+                if (numMembersWithExtraPartition > 0) {
+                    remaining++;
+                    numMembersWithExtraPartition--;
+                }
+                if (remaining > 0) {
+                    RemainingAssignmentsForMember<String, Integer> newPair = new RemainingAssignmentsForMember<>(memberId, remaining);
+                    unfilledMembersPerTopic
+                        .computeIfAbsent(topicId, k -> new ArrayList<>())
+                        .add(newPair);
+                }
+            }
+        });
+
+        // Step 4
+        // Find the difference between the total partitions per topic and the already assigned sticky partitions for the topic to get unassigned partitions.
+        Map<Uuid, List<Integer>> unassignedPartitionsPerTopic = getUnassignedPartitionsPerTopic(assignmentSpec, assignedStickyPartitionsPerTopic);
+
+        // Step 5
+        // Iterate through the unfilled members list and assign the remaining number of partitions from the unassignedPartitions list.
+        unfilledMembersPerTopic.forEach((topicId, unfilledMembersForTopic) -> {
+            int unassignedPartitionsListStartPointer = 0;
+            for (RemainingAssignmentsForMember<String, Integer> remainingAssignmentsForMember : unfilledMembersForTopic) {
+                String memberId = remainingAssignmentsForMember.memberId();
+                int remaining = remainingAssignmentsForMember.remaining();
+                List<Integer> partitionsToAssign = unassignedPartitionsPerTopic.get(topicId)
+                    .subList(unassignedPartitionsListStartPointer, unassignedPartitionsListStartPointer + remaining);
+
+                unassignedPartitionsListStartPointer += remaining;
+                membersWithNewAssignmentPerTopic
+                    .computeIfAbsent(memberId, k -> new HashMap<>())
+                    .computeIfAbsent(topicId, k -> new HashSet<>())
+                    .addAll(partitionsToAssign);
+            }
```

Comment:
Should we break this loop when there are no more partitions left to be assigned?

Reply:
Theoretically the sum of all the "remaining" values in the unfilled members list for the topic will be equal to the total unassigned partitions so we don't need to break the loop cause it happens automatically. I could add a check to ensure this is the case, I've added a check in the uniform assignor anyways just for a correctness check.